components/electric/lib/electric/postgres/schema/fk_graph.ex

```diff
@@ -10,40 +10,82 @@ defmodule Electric.Postgres.Schema.FkGraph do
   key relations in a separate map of `%{relation() => %{relation() => [column_name()]}}`.
   """
   alias Electric.Postgres.Schema.Proto
+  alias Electric.Postgres.Extension.SchemaLoader
 
-  defstruct [:graph, fks: %{}]
+  defstruct [:directed, :undirected, fks: %{}, pks: %{}]
 
+  @type relation() :: Electric.Postgres.relation()
+  @type name() :: Electric.Postgres.name()
+  @type fks() :: [name(), ...]
+  @type pks() :: [name(), ...]
+  @type join() ::
+          {:many_to_one, {relation(), fks()}, {relation(), pks()}}
+          | {:one_to_many, {relation(), pks()}, {relation(), fks()}}
+
+  @type t() :: %__MODULE__{
+          directed: Graph.t(),
+          undirected: Graph.t(),
+          fks: %{relation() => %{relation() => fks()}},
+          pks: %{relation() => pks()}
+        }
+  @type edge() :: {relation(), relation(), label: fks()} | {relation(), relation(), fks()}
+  @type path() :: [relation(), ...]
+
+  @spec for_schema(SchemaLoader.Version.t()) :: t()
+  def for_schema(%SchemaLoader.Version{schema: schema}) do
+    for_schema(schema)
+  end
+
+  @spec for_schema(%Proto.Schema{}) :: t()
   def for_schema(%Proto.Schema{tables: tables}) do
-    tables
-    |> Stream.flat_map(fn %Proto.Table{constraints: constraints, name: name} ->
-      constraints
-      |> Stream.filter(&match?(%{constraint: {:foreign, _}}, &1))
-      |> Enum.map(fn %{constraint: {:foreign, fk}} ->
-        {{name.schema, name.name}, {fk.pk_table.schema, fk.pk_table.name}, fk.fk_cols}
+    fks =
+      tables
+      |> Enum.flat_map(fn %Proto.Table{constraints: constraints, name: name} ->
+        constraints
+        |> Stream.filter(&match?(%{constraint: {:foreign, _}}, &1))
+        |> Enum.map(fn %{constraint: {:foreign, fk}} ->
+          {{name.schema, name.name}, {fk.pk_table.schema, fk.pk_table.name}, fk.fk_cols}
+        end)
       end)
```

Comment:
Merely a code style suggestion. I believe this is simpler to parse when reading the code:

```suggestion
      for %Proto.Table{constraints: constraints, name: name} <- tables,
          %{constraint: {:foreign, fk}} <- constraints do
        {{name.schema, name.name}, {fk.pk_table.schema, fk.pk_table.name}, fk.fk_cols}
      end
```

Reply:
i never reach for comprehensions, but you have a point